controlplane/modules/route/internal/rib/map_trie.go

```diff
@@ -0,0 +1,70 @@
+package rib
+
+import (
+	"net/netip"
+)
+
+type MapTrieKey struct {
+	Prefix netip.Prefix
+}
+
+type MapTrie [129]map[MapTrieKey]*RoutesList
+
+func NewMapTrie(preAllocSize int) MapTrie {
+	mt := MapTrie{}
+	for idx := range mt {
+		mt[idx] = make(map[MapTrieKey]*RoutesList, preAllocSize)
+	}
+	return mt
+}
+
+func (m *MapTrie) Lookup(addr netip.Addr) (*RoutesList, bool) {
+	maxBits := 129
+	base := 128
+	if addr.Is4() {
+		base = 32
+		maxBits = 33
+	}
+	for n := range maxBits {
+		bits := base - n
+		p, err := addr.Prefix(bits)
+		if err != nil {
+			panic("Imposible err: " + err.Error())
+		}
+		mtk := MapTrieKey{Prefix: p}
+		if _, ok := m[bits][mtk]; ok {
+			return nil, false
+		}
+	}
+	return nil, false
+}
+
+func (m *MapTrie) Insert(key MapTrieKey, route Route) {
```

Comment:
Suggest instead of just `Insert` function like `InsertOrUpdate(key, onVacant func() Route, onOccupied func(*Route))`

Reply:
Or [entry-like](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html) API